storage/redis/peer_store.go

```diff
@@ -0,0 +1,762 @@
+		rb.redsync = redsync.New(pools)
+	}
+	return rb.pool.Get()
+}
+
+// New creates a new PeerStore backed by redis.
+func New(provided Config) (storage.PeerStore, error) {
+	cfg := provided.Validate()
+
+	// creates Backend instance
+	u, err := common.ParseRedisURL(cfg.RedisBroker)
+	if err != nil {
+		return nil, err
+	}
+
+	ps := &peerStore{
+		cfg:    cfg,
+		rb:     NewBackend(u.Host, u.Password, "", u.DB),
+		closed: make(chan struct{}),
+	}
+
+	// Start a goroutine for garbage collection.
+	ps.wg.Add(1)
+	go func() {
+		defer ps.wg.Done()
+		for {
+			select {
+			case <-ps.closed:
+				return
+			case <-time.After(cfg.GarbageCollectionInterval):
+				before := time.Now().Add(-cfg.PeerLifetime)
+				log.Debug("storage: purging peers with no announces since", log.Fields{"before": before})
+				ps.collectGarbage(before)
+			}
+		}
+	}()
+
+	// Start a goroutine for reporting statistics to Prometheus.
+	ps.wg.Add(1)
+	go func() {
+		defer ps.wg.Done()
+		t := time.NewTicker(cfg.PrometheusReportingInterval)
+		for {
+			select {
+			case <-ps.closed:
+				t.Stop()
+				return
+			case <-t.C:
+				before := time.Now()
+				ps.populateProm()
+				log.Debug("storage: populateProm() finished", log.Fields{"timeTaken": time.Since(before)})
+			}
+		}
+	}()
+
+	return ps, nil
+}
+
+type serializedPeer string
+
+func newPeerKey(p bittorrent.Peer) serializedPeer {
+	b := make([]byte, 20+2+len(p.IP.IP))
+	copy(b[:20], p.ID[:])
+	binary.BigEndian.PutUint16(b[20:22], p.Port)
+	copy(b[22:], p.IP.IP)
+
+	return serializedPeer(b)
+}
+
+func decodePeerKey(pk serializedPeer) bittorrent.Peer {
+	peer := bittorrent.Peer{
+		ID:   bittorrent.PeerIDFromString(string(pk[:20])),
+		Port: binary.BigEndian.Uint16([]byte(pk[20:22])),
+		IP:   bittorrent.IP{IP: net.IP(pk[22:])}}
+
+	if ip := peer.IP.To4(); ip != nil {
+		peer.IP.IP = ip
+		peer.IP.AddressFamily = bittorrent.IPv4
+	} else if len(peer.IP.IP) == net.IPv6len { // implies toReturn.IP.To4() == nil
+		peer.IP.AddressFamily = bittorrent.IPv6
+	} else {
+		panic("IP is neither v4 nor v6")
+	}
+
+	return peer
+}
+
+type peerStore struct {
+	cfg Config
+	rb  *Backend
+
+	closed chan struct{}
+	wg     sync.WaitGroup
+}
+
+// populateProm aggregates metrics over all groups and then posts them to
+// prometheus.
+func (ps *peerStore) populateProm() {
+	var numInfohashes, numSeeders, numLeechers uint64
+
+	groups := [2]string{bittorrent.IPv4.String(), bittorrent.IPv6.String()}
+
+	conn := ps.rb.open()
+	defer conn.Close()
+
+	for _, group := range groups {
+		ihList, err := conn.Do("HKEYS", group) // key
+		if err != nil {
+			log.Error("storage: execute redis HKEYS error", log.Fields{
+				"Hkey":  group,
+				"Error": err.Error(),
+			})
+			return
+		}
+		infohashes := ihList.([]interface{})
+
+		infohashLPrefix := group + "_L_"
+		infohashSPrefix := group + "_S_"
+		infohashPrefixLen := len(infohashLPrefix)
+		infohashesMap := make(map[string]struct{})
+
+		for _, ihKey := range infohashes {
```

Comment:
Is this going to scale? It's probably better see if we can leverage `COUNT` in some way or keep our own statistics in redis and use transactions to update those keys with every change to peers (which could be tricky if we use `EXPIRE`).

Reply:
A peer has multiple infohashes, incr/decr redis counter on putting/deleting seeder/leecher is not  exactly what we expect, leecher and seeder set are needed which is more time-consuming

Reply:
> A peer has multiple infohashes

We collect at least these two metrics in prometheus:
- The total number of infohashes
- The total number of peers across all infohashes

We don't collect the number of _distinct_ peers (i.e. the number of _clients_) across all infohashes, as that'd require tremendous amounts of memory or time. If a user is running a _client_ that is a _peer_ in multiple swarms, that's fine, each peer counts for prometheus.

I like the idea of using a counter in redis, but we'd have to make it so that it's actually correct, i.e. a put with the peer already present does not increase the counter. Not sure how simple that is, I'm not doing redis much. Alternatively, is there not something that allows us to range over all the fields in a hash, but without getting them all from redis in one go? Or does that block something? Again, not an expert.

On another note: Looking at this method, I don't think the calculated numLeechers and numSeeders are correct. Neither do I think the number of infohashes is correct. Let me illustrate what I mean:

As far as I understand, redis looks something like this:
```
- IPv4
  - IPv4_S_<infohash 1>: <modification time>
  - IPv4_L_<infohash 1>: <modification time>
  - IPv4_S_<infohash 2>: <modification time>
- IPv4_S_<infohash 1>
  - <peer 1 key>: <modification time>
  - <peer 2 key>: <modification time>
- IPv4_L_<infohash 1>
  - <peer 3 key>: <modification time>
- IPv4_S_<infohash 2>
  - <peer 3 key>: <modification time>
```
**In this case, prometheus should record two swarms and four peers.**

Now, you iterate over each key under `IPv4` and put them in the `infohashesMap`. You count that map to get the number of infohashes/swarms. This is not correct, as you count an infohash with both seeders and leechers as two.

You then also check whether the prefix is `IPv4_S_` or `IPv4_L_` and, depending on the outcome, count that as one seeder or one leecher. This is not correct, as the swarms may have more than one peer in them.

Instead:
- Get all fields under `IPv4` key, lets call them `prefixedInfohashes`
- Split each `prefixedInfohash` of them into `prefix` and `infohash`
- Collect the `infohash` part in a map (I'd call this `swarms`, to avoid confusion), at the end get the count of the map for the number of swarms
- Ask redis about the number `n` of fields under the `prefixedInfohash` key. If the `prefix` indicates seeders, add `n` to the total number of seeders. Vice versa for leechers.
- Repeat for IPv6

Reply:
You are right, the method is incorrect.

Reply:
There is a question now, redis is a centralized storage, each chihaya reports prometheus in its own period, which is incorrect.

Reply:
We can use redis command `hset` to check if the peer exist present, but we could not optimize the `swarms` metric. I think we can make a compromise to improve performance, to take `IPv4_infohash_count` as seeder's infohash only, or split the `IPv4_infohash_count` to two metrics `IPv4_infohash_seeder_count` and `IPv4_infohash_leecher_count`.